server/resp/src/main/java/org/infinispan/server/resp/commands/string/MGET.java

```diff
@@ -41,28 +43,47 @@ public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
       for (int i = 0; i < keysToRetrieve; ++i) {
          int innerCount = i;
          byte[] keyBytes = arguments.get(i);
-         getStage.dependsOn(handler.cache().getAsync(keyBytes)
-               .whenComplete((returnValue, t) -> {
-                  if (returnValue != null) {
-                     results.set(innerCount, returnValue);
-                     int length = returnValue.length;
-                     if (length > 0) {
-                        // $ + digit length (log10 + 1) + /r/n + byte length
-                        resultBytesSize.addAndGet(1 + (int) Math.log10(length) + 1 + 2 + returnValue.length);
-                     } else {
-                        // $0 + /r/n
-                        resultBytesSize.addAndGet(2 + 2);
-                     }
-                  } else {
-                     // $-1
-                     resultBytesSize.addAndGet(3);
-                  }
-                  // /r/n
-                  resultBytesSize.addAndGet(2);
-               }));
+         CompletableFuture<byte[]> async;
+         try {
+            async = handler.cache().getAsync(keyBytes).handle((v, t) -> {
+               if (t != null)
+                  throw new RuntimeException(t);
+               return v;
```

Comment:
this handle seems superfluous, but I'm unable to make it works without it. tests passes but server fails with resp client.
@jabolina please have a look when you have time

Reply:
I scratched my head with this. But what happens is that we load the entry correctly, and the `getAsync` future completes successfully. The type check executes when we have something subscribed, so the cast fails internally. Calling the first `.handle` fails to convert from the object to `byte[]` and leaves the CF as failed exceptionally, which is caught on the handle below, treating the exception correctly. I made some changes and ended with the following:

```diff
diff --git a/server/resp/src/main/java/org/infinispan/server/resp/commands/string/MGET.java b/server/resp/src/main/java/org/infinispan/server/resp/commands/string/MGET.java
index 51122c4a05..898d2202b4 100644
--- a/server/resp/src/main/java/org/infinispan/server/resp/commands/string/MGET.java
+++ b/server/resp/src/main/java/org/infinispan/server/resp/commands/string/MGET.java
@@ -1,20 +1,20 @@
 package org.infinispan.server.resp.commands.string;
 
-import java.util.Arrays;
-import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.commons.util.concurrent.CompletionStages;
 import org.infinispan.server.resp.ByteBufferUtils;
 import org.infinispan.server.resp.Resp3Handler;
 import org.infinispan.server.resp.RespCommand;
 import org.infinispan.server.resp.RespErrorUtil;
 import org.infinispan.server.resp.RespRequestHandler;
 import org.infinispan.server.resp.commands.Resp3Command;
-import org.infinispan.commons.util.concurrent.AggregateCompletionStage;
-import org.infinispan.commons.util.concurrent.CompletionStages;
 
 import io.netty.channel.ChannelHandlerContext;
 
@@ -23,6 +23,13 @@
  * @since 14.0
  */
 public class MGET extends RespCommand implements Resp3Command {
+
+   private static final Function<Object, byte[]> TYPE_CHECKER = res -> {
+      if (!(res instanceof byte[]))
+         return null;
+      return (byte[]) res;
+   };
+
    public MGET() {
       super(-2, 1, -1, 1);
    }
@@ -36,54 +43,44 @@ public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
          ByteBufferUtils.stringToByteBufAscii("*0\r\n", handler.allocator());
          return handler.myStage();
       }
-      List<byte[]> results = Collections.synchronizedList(Arrays.asList(
-            new byte[keysToRetrieve][]));
       AtomicInteger resultBytesSize = new AtomicInteger();
-      AggregateCompletionStage<Void> getStage = CompletionStages.aggregateCompletionStage();
-      for (int i = 0; i < keysToRetrieve; ++i) {
-         int innerCount = i;
-         byte[] keyBytes = arguments.get(i);
-         CompletableFuture<byte[]> async;
-         try {
-            async = handler.cache().getAsync(keyBytes).handle((v, t) -> {
-               if (t != null)
-                  throw new RuntimeException(t);
-               return v;
-            });
-         } catch (Exception ex) {
-            async = CompletableFuture.completedFuture(handleWrongTypeError(ex));
-         }
-         getStage.dependsOn(async.handle((returnValue, t) -> {
-            if (t != null) {
-               returnValue = handleWrongTypeError(t);
-            }
-            results.set(innerCount, returnValue);
-            if (returnValue != null) {
-               int length = returnValue.length;
-               if (length > 0) {
-                  // $ + digit length (log10 + 1) + /r/n + byte length
-                  resultBytesSize.addAndGet(1 + (int) Math.log10(length) + 1 + 2 + returnValue.length);
+      CompletionStage<List<byte[]>> results = CompletionStages.performSequentially(arguments.iterator(), k -> getAsync(handler, k)
+            .exceptionally(MGET::handleWrongTypeError)
+            .handle((returnValue, t) -> {
+               if (returnValue != null) {
+                  int length = returnValue.length;
+                  if (length > 0) {
+                     // $ + digit length (log10 + 1) + /r/n + byte length
+                     resultBytesSize.addAndGet(1 + (int) Math.log10(length) + 1 + 2 + returnValue.length);
+                  } else {
+                     // $0 + /r/n
+                     resultBytesSize.addAndGet(2 + 2);
+                  }
                } else {
-                  // $0 + /r/n
-                  resultBytesSize.addAndGet(2 + 2);
+                  // $-1
+                  resultBytesSize.addAndGet(3);
                }
-            } else {
-               // $-1
-               resultBytesSize.addAndGet(3);
-            }
-            // /r/n
-            resultBytesSize.addAndGet(2);
-            return null;
-         }));
+               // /r/n
+               resultBytesSize.addAndGet(2);
+               return returnValue;
+            }), Collectors.toList());
+      return handler.stageToReturn(results, ctx, (r, alloc) -> ByteBufferUtils.bytesToResult(resultBytesSize.get(), r, alloc));
+   }
+
+   private static CompletionStage<byte[]> getAsync(Resp3Handler handler, byte[] key) {
+      CompletableFuture<?> async;
+      try {
+         async = handler.cache().getAsync(key);
+      } catch (Exception ex) {
+         async = CompletableFuture.completedFuture(handleWrongTypeError(ex));
       }
-      return handler.stageToReturn(getStage.freeze(), ctx,
-            (ignore, alloc) -> ByteBufferUtils.bytesToResult(resultBytesSize.get(), results, alloc));
+      return async.thenApply(TYPE_CHECKER);
    }
 
    private static byte[] handleWrongTypeError(Throwable ex) {
       if (RespErrorUtil.isWrongTypeError(ex)) {
          return null;
       }
-      throw (RuntimeException) ex;
+      throw CompletableFutures.asCompletionException(ex);
    }
 }
```

Reply:
thanks @jabolina , I imported your changes